app/controllers/uploads_controller.rb

```diff
@@ -270,6 +301,186 @@ def complete_external_upload
+      presigned_urls[part_number] = Discourse.store.presign_multipart_part(
+        upload_id: external_upload_stub.external_upload_identifier,
+        key: external_upload_stub.key,
+        part_number: part_number
+      )
+    end
+
+    render json: { presigned_urls: presigned_urls }
+  end
+
+  def validate_part_number(part_number)
+    if part_number.zero? || part_number.negative? || !part_number.between?(1, 10000)
+      raise Discourse::InvalidParameters.new(
+        "Part numbers should be a list of numbers between 1 and 10000"
+      )
+    end
+  end
+
+  def multipart_upload_exists?(external_upload_stub)
+    begin
+      Discourse.store.list_multipart_parts(
+        upload_id: external_upload_stub.external_upload_identifier, key: external_upload_stub.key
+      )
+    rescue Aws::S3::Errors::NoSuchUpload => err
+      debug_upload_error(err, "upload.external_upload_not_found", { additional_detail: "path: #{external_upload_stub.key}" })
+      return false
+    end
+    true
+  end
+
+  def abort_multipart
+    return render_404 if !SiteSetting.enable_direct_s3_uploads
+
+    external_upload_identifier = params.require(:external_upload_identifier)
+
+    RateLimiter.new(
+      current_user, "abort-multipart-upload", ABORT_MULTIPART_RATE_LIMIT_PER_MINUTE, 1.minute
+    ).performed!
+
+    external_upload_stub = ExternalUploadStub.find_by(
+      external_upload_identifier: external_upload_identifier,
+      created_by: current_user
+    )
+    return render_404 if external_upload_stub.blank?
+
+    begin
+      abort_response = Discourse.store.abort_multipart(
+        upload_id: external_upload_stub.external_upload_identifier,
+        key: external_upload_stub.key
+      )
+    rescue Aws::S3::Errors::ServiceError => err
+      debug_upload_error(err, "upload.abort_mutlipart_failure", additional_detail: "external upload stub id: #{external_upload_stub.id}")
+      return render_json_error(I18n.t("upload.failed"), status: 422)
+    end
+
+    external_upload_stub.destroy!
+
+    render json: success_json
+  end
+
+  def complete_multipart
+    return render_404 if !SiteSetting.enable_direct_s3_uploads
+
+    unique_identifier = params.require(:unique_identifier)
+    parts = params.require(:parts)
+
+    RateLimiter.new(
+      current_user, "complete-multipart-upload", COMPLETE_MULTIPART_RATE_LIMIT_PER_MINUTE, 1.minute
+    ).performed!
+
+    external_upload_stub = ExternalUploadStub.find_by(
+      unique_identifier: unique_identifier, created_by: current_user
+    )
+    return render_404 if external_upload_stub.blank?
+
+    if !multipart_upload_exists?(external_upload_stub)
+      return render_404
+    end
+
+    parts = parts.map do |part|
+      part_number = part[:part_number].to_i
+      etag = part[:etag]
+      validate_part_number(part_number)
+
+      if part_number.blank? || etag.blank?
+        raise Discourse::InvalidParameters.new("All parts must have a part number between 1 and 10000 and an ETag")
```

Comment:
Not sure about mentioning part number here since it should have been caught by `validate_part_number`

Reply:
Yes true I can just check etag